File: Year II/First Semester/MAP/Seminar/seminar1/seminar1/src/decorator/StrategyTaskRunner.java
package decorator;

import factory.ContainerFactory;
import factory.Strategy;
import model.Container;
import model.Task;

public class StrategyTaskRunner implements TaskRunner {
    private Container container;

    public StrategyTaskRunner(Strategy strategy) {
        ContainerFactory containerFactory = new ContainerFactory();
        container = containerFactory.createContainer(strategy);
    }

    @Override
    public void executeOneTask() {
        if (hasTask()) {
            Task removedTask = container.remove();
            removedTask.execute();
        }
        else {
            System.out.println("There are no tasks!");
        }
    }

    @Override
    public void executeAll() {
        while (hasTask()) {
            executeOneTask();
        }
    }

    @Override
    public void addTask(Task t) {
        container.add(t);
    }

    @Override
    public boolean hasTask() {
        return !container.isEmpty();
    }
}

File: Year II/First Semester/MAP/Laborator/Laborator9/ConsoleApp1/ConsoleApp1/Decorator/StrategyTaskRunner.cs
using ConsoleApp1.Factory;
using ConsoleApp1.Model;
using Task = ConsoleApp1.Model.Task;

namespace ConsoleApp1.Decorator;

public class StrategyTaskRunner: TaskRunner
{
    private readonly Container _container;

    public StrategyTaskRunner(Strategy strategy)
    {
        _container = TaskContainerFactory.FactoryInstance.createContainer(strategy);
    }

    public void AddTask(Task task)
    {
        _container.Add(task);
    }

    public bool HasTasks()
    {
        return !(_container.Empty());
    }

    public void ExecuteOneTask()
    {
        _container.Remove().Execute();
    }

    public void ExecuteAll()
    {
        while (HasTasks())
            ExecuteOneTask();
    }
}